Actions cheatsheet (#692)
github/training-kit: downloads/github-actions-cheat-sheet.md

---
layout: cheat-sheet
redirect_to: false
title: GitHub Actions Cheat Sheet

> byline: GitHub Actions give you the flexibility to build automated software development lifecycle workflows. You can write individual tasks, called actions, and combine them to create  custom workflows in your repository. GitHub Actions are automated processes allowing you to build, test, package, release, or deploy any code project on GitHub, but you can also use them to automate any step of your workflow&#58; merging pull requests, assigning labels, triaging issues to name a few.

Reviewer: ```suggestion
byline: GitHub Actions give you the flexibility to build automated software development lifecycle workflows. You can write individual tasks, called actions, and combine them to create custom workflows in your repository. GitHub Actions are automated processes allowing you to build, test, package, release, or deploy any code project on GitHub, but you can also use them to automate any step of your workflow&#58; merging pull requests, assigning labels, triaging issues to name a few.
```

Removed extra whitespace, but what about the `workflow&#58`, I think the `;` is already there. Should we take it out?

Reviewer 2: ":" is a Jekyll delelimiter so I had to use the HTML code

leadingpath: ../../
---

{% capture colOne %}
## Workflow Syntax

> Workflow files use YAML syntax, and must have either a .yml or .yaml file extension. You must store workflow files in the `.github/workflows/` directory of your repository.

Reviewer: ```suggestion
Workflow files use YAML syntax, and must have either a .yml or .yaml file extension. You must store workflow files in the `.github/workflows/` directory of your repository. Each different YAML file corresponds to a different workflow.
```

```yaml
name: Node CI
on: push
  branches:    
    - 'releases/*'
jobs:
  my_build:
    runs-on: ubuntu-latest
    needs: my-other-job
    steps:
    - uses: actions/checkout@master
    - name: Say something
      run: echo "A little bit more action"
```

### Workflow `name`

> The name of your workflow. GitHub displays the names of your workflows on your repository's actions page.

Reviewer 3: Are there any naming rules? Forbidden characters?  Nothing in the docs but I wonder if special chars are forbidden.

> The name of your workflow. GitHub displays the names of your workflows on your repository's actions page.

Reviewer 3: ```suggestion
The name of your workflow. GitHub displays the names of your workflows on your repository's actions page.  Defaults to your workflow filename if omitted.
```

Reviewer 2: We are aiming for a 2 pager so we need to balance completness with real estate optmization.

### `on` Event
The name of the GitHub event that triggers the workflow. You can provide a single event string, an array of events, or an event configuration map that restricts the execution of a workflow:
- When using the `push` and `pull_request` events, `branches` and `tags` allow to select or exclude references the workflow will run on, while `paths` specifies which files must have been modified in order to run the workflow.  
- The `types` keyword enables you to narrow down activity that causes the workflow to run: `created`, `edited`, `deleted`...
A workflow trigger can also be scheduled:  
```yaml

> on:

Reviewer 3: The schedule example here seems like a non-sequitur since it isn't mentioned in the text above.  Perhaps include an example mentioned therein.  Also, I think it's worth highlighting that you can filter by `path` as well as `paths-ignore` (although not in the same job) . 

If we mention the `schedule` option, consider adding a link to the [syntax explanation](https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#onschedule)

Reviewer 2: Absolutely. Something disappeard here and new options have been released since it was written. The whole section need a refresh.

  schedule:
  - cron: '*/15 * * * *'
```

### `jobs` Collection
A workflow run is made up of one or more jobs identified by a `job_id`. Jobs  run in parallel by default. Each job runs in a fresh instance of the virtual environment specified by `runs-on`.

#### `job_id`
The name of the job displayed on GitHub. For example `my_build` or `build`. This string is unique with the `jobs` object.

{% endcapture %}
<div class="col-md-6">
{{ colOne | markdownify }}
</div>

{% capture colTwo %}

#### `needs`
Identifies any job that must complete successfully before this job will run. It can be a string or array of strings. If a job fails, all jobs that need it are skipped unless the jobs use a conditional statement that causes the job to continue.

#### `runs-on`

> Inside of the `jobs` collection, this is the type of virtual host machine to run the job on.

Reviewer 3: ```suggestion
The type of virtual host machine to run the job on.  Can be either a GitHub or self-hosted runner.  Jobs can also run in user-specified containers (see: `container`)
```

Reviewer: I was going to edit the `inside of the job` but I see @bryancross already proposed a change so :+1:

      options: --cpus 1
```

#### `services`
Additional containers to host services for a job in a workflow. These are useful for creating databases or cache services. The runner on the virtual machine will automatically create a network and manage the lifecycle of the service containers. Each service is a named object in the `services` collection (`redis` or `nginx` for example) and can receive the same parameters than the `container` object.

#### `timeout-minutes`
The maximum number of minutes to let a workflow run before GitHub automatically cancels it. Default: 360

{% endcapture %}
<div class="col-md-6">
{{ colTwo | markdownify }}
</div>
<div class="clearfix"></div>

{% capture colThree %}
### Job `steps`
A job contains a sequence of tasks called `steps`. Steps can run commands, run setup tasks, or run an action from your repository, a public repository, or an action published in a Docker registry. Each step runs in its own process in the virtual environment and has access to the workspace and filesystem.

#### Step `name`
Specify the label to be displayed for this step in GitHub. It's not required but does improve readability in the logs.

#### `uses`

> Select an action to run as part of a step in your job. You can use an action defined in the same repository as the workflow, a public repository elsewhere on GitHub, or in a published Docker container image.

Reviewer 3: ```suggestion
Specify an action to run as part of a step in your job. You can use an action defined in the same repository as the workflow, a public repository elsewhere on GitHub, or in a published Docker container image.
```

#### `with`

> A map of the input parameters defined by the action. Each input parameter is a key/value pair. Input parameters are prefixed with `INPUT_`, converted to upper case and set as environment variables: the `first_name` parameter will be exposed as the `INPUT_FIRST_NAME` environment variable. Special parameter names are:

Reviewer 3: ```suggestion
A map of the input parameters defined by the action in its `action.yml` file. Each input parameter is a key/value pair. Input parameters are prefixed with `INPUT_`, converted to upper case and set as environment variables: the `first_name` parameter will be exposed as the `INPUT_FIRST_NAME` environment variable. Special parameter names are:
```

- `args`, a string that defines the inputs passed to a Docker container's `ENTRYPOINT`. It is used in place of the `CMD` instruction in a `Dockerfile`.
- `entrypoint`, a string that defines or overrides the executable to run as the Docker container's `ENTRYPOINT`.

#### `env`

> Sets environment variables for steps to use in the virtual environment.  If you are setting a secret in an environment variable, you must use the `secrets` context.

Reviewer 3: A word on whether this is a safe practice would be helpful here.

Reviewer 2: Not sure what to say here and whether it fits on a cheat sheet

```yaml
...
steps:
  - name: A cool action
    uses: actions/bin/sh@master
    with:
      entrypoint: /bin/echo
      args: The ${{ github.event_name }} did it
    env:
      last_name: ${{secrets.LASTNAME}}
```

> Default environment variables:

Reviewer 3: Missing `GITHUB_ACTIONS`

Reviewer 2: I'm thinking about shrinking this section and keeping essential variables only

`HOME`: Path to the GitHub home directory used to store user data.
`GITHUB_WORKFLOW`: Name of the workflow.
`GITHUB_ACTION`: Name of the action.
`GITHUB_ACTOR`: Name of the person or app that initiated the workflow.
`GITHUB_REPOSITORY`: Owner and repository name.
`GITHUB_EVENT_NAME`: Name of the triggering webhook event.
`GITHUB_EVENT_PATH`: Path of the file with the webhook event payload.
`GITHUB_WORKSPACE`: GitHub workspace directory path.
`GITHUB_SHA`: Commit SHA that triggered the workflow.
`GITHUB_REF`: Branch or tag ref that triggered the workflow.
`GITHUB_HEAD_REF`: Branch of the head repository (forks only).
`GITHUB_BASE_REF`: Branch of the base repository (forks only).

> #### `run`

Reviewer 3: can run multiple commands in a single shell instance using `|`

## Syntax
Workflow files use YAML syntax, and must have either a .yml or .yaml file extension. You must store workflow files in the `.github/workflows` directory of your repository.

### Workflow `name`
The name of your workflow. GitHub displays the names of your workflows on your repository's actions page.

### `on` Event
Required The name of the GitHub event that triggers the workflow. You can provide a single event string, array of events, or an event configuration map that schedules a workflow or restricts the execution of a workflow to specific files, tags, or branch changes.

```
# Trigger on push
on: push
```

```
# Trigger the workflow on push or pull request
on: [push, pull_request]
```

{% endcapture %}
<div class="col-md-6">
{{ colOne | markdownify }}
</div>

{% capture colTwo %}

### `jobs`
A workflow run is made up of one or more jobs. Jobs run in parallel by default. To run jobs sequentially, you can define dependencies on other jobs using the `jobs.<job_id>.needs` keyword. Each job runs in a fresh instance of the virtual environment specified by runs-on.

> ### `runs-on`

Author: One thing we may want to do with commands after our jobs heading is make them level 4 headers #### because it continues the hierarchy that is set up in the yaml. runs-on for instance is actually jobs.<job_id>.runs-on in the documentation.

Thoughts?

Author: it ends up looking nicer with `<h4>` headings. marking this as resolved.

Inside of the `jobs` collection, this is the type of virtual host machine to run the job on. Each job runs with a fresh instance of the virtual environment specified in by `runs-on`.

Available virtual machine types are:

- `ubuntu-latest`, `ubuntu-18.04`, or `ubuntu-16.04`
- `windows-latest`, `windows-2019`, or `windows-2016`

> - `macOS-latest` or `macOS-10.14`

Author: I'm a bit apprehensive for adding versions here because this makes futureproofing it real difficult. I'd like to show something like just the latest versions.

Reviewer 3: Agreed, include a link to the relevant docs section for latest info
